src/extensions/default/SVGCodeHints/main.js

```diff
@@ -0,0 +1,273 @@
+            });
+        }
+        return cachedAttributes[tagName];
+    }
+    
+    PreferencesManager.on("change", "codehint.SVGHints", function () {
+        isSVGEnabled = _isSVGHintsEnabled();
+    });
+    
+    PreferencesManager.on("change", "showCodeHints", function () {
+        isSVGEnabled = _isSVGHintsEnabled();
+    });
+    
+    // Check if SVG Hints are available.
+    isSVGEnabled = _isSVGHintsEnabled();
+    
+    /**
+     * @constructor
+     */
+    function SVGCodeHints() {
+        this.tagInfo = null;
+    }
+    
+    /**
+     * Determines whether SVG code hints are available in the current editor.
+     * 
+     * @param {!Editor} editor An instance of Editor
+     * @param {string} implicitChar A single character that was inserted by the
+     * user or null if the request was explicity made to start hinting session.
+     * 
+     * @return {boolean} Determines whether or not hints are available in the current context.
+     */
+    SVGCodeHints.prototype.hasHints = function (editor, implicitChar) {
+        if (isSVGEnabled && editor.getModeForSelection() === "image/svg+xml") {
+            this.editor = editor;
+            this.tagInfo = XMLUtils.getTagInfo(this.editor, this.editor.getCursorPos());
+            
+            if (this.tagInfo && this.tagInfo.tokenType) {
+                return true;
+            }
+        }
+        return false;
+    };
+    
+    /**
+     * Returns a list of hints that are available in the current context,
+     * or null if there are no hints available. 
+     *
+     * @param {string} implicitChar A character that the user typed in the hinting session.
+     * @return {!{hints: Array.<string>, match: string, selectInitial: boolean, handleWideResults: boolean}}
+     */
+    SVGCodeHints.prototype.getHints = function (implicitChar) {
+        var hints = [], query, tagInfo, i, attrs = [], index, isMultiple, tagSpecificAttrs;
+        
+        tagInfo  = XMLUtils.getTagInfo(this.editor, this.editor.getCursorPos());
+        this.tagInfo = tagInfo;
+        
+        if (tagInfo && tagInfo.tokenType) {
+            query = tagInfo.token.string.substr(0, tagInfo.offset).trim();
+        
+            if (tagInfo.tokenType === XMLUtils.TOKEN_TAG) {
+                for (i in tagData.tags) {
+                    if (tagData.tags.hasOwnProperty(i) && i.indexOf(query) === 0) {
+                        hints.push(i);
+                    }
+                }
+            } else if (tagInfo.tokenType === XMLUtils.TOKEN_ATTR) {
+                if (!tagData.tags[tagInfo.tagName]) {
+                    return null;
+                }
+                // Get attributes.
+                attrs = getTagAttributes(tagInfo.tagName);
+                hints = attrs.filter(function (value) {
+                    if (value.indexOf(query) === 0 && tagInfo.exclusionList.indexOf(value) === -1) {
+                        return true;
+                    }
+                });
```

Comment:
For code hints, we use the `StringMatch` object for smarter filtering. For example, in CSS type `position: sti` and you'll notice that in addition to showing `sticky`, `static` is also in list since is has the all of the letters `s`, `t`, and `i` in order, but not contiguous.

Take a look at CSSCodeHints/main.js to see how it uses `StringMatch`.